internal/provider/aws/provider_test.go

```diff
@@ -0,0 +1,4906 @@
+								types.Tag{
+									Key:   aws.String(string(TagKeyExternalIPName)),
+									Value: aws.String(externalIP.Name),
+								},
+								types.Tag{
+									Key:   aws.String(string(TagKeyInstanceID)),
+									Value: aws.String(instanceID),
+								},
+							))
+
+							return &ec2.AllocateAddressOutput{
+								AllocationId: aws.String(allocationID),
+							}, nil
+						})
+
+					status, err := provider.ReconcileExternalIP(ctx, log, instanceID, externalIP)
+					Expect(status).To(MatchFields(IgnoreExtras, Fields{
+						"State": Equal(v1alpha1.ExternalIPStateReserved),
+						"Conditions": ConsistOf(matchConditions([]kmetav1.Condition{
+							{
+								Type:   v1alpha1.ExternalIPConditionReasonIPCreated,
+								Status: kmetav1.ConditionTrue,
+								Reason: v1alpha1.ExternalIPConditionReasonIPCreated,
+							},
+							{
+								Type:   v1alpha1.ExternalIPConditionTypeNetworkInterfaceAssociated,
+								Status: kmetav1.ConditionFalse,
+								Reason: v1alpha1.ExternalIPConditionReasonProviderError,
+							},
+						}, "LastTransitionTime", "ObservedGeneration", "Message")),
+					}))
+					Expect(err).ToNot(HaveOccurred())
+				})
+			})
+
+			When("the address is associated to an instance", func() {
+				associationID = "eipassoc-" + testID
+
+				It("should return a reserved state and specify IPCreation and NetworkInterfaceAssociation condition and an error when an AWS API call (DisassociateAddress) fails", func() {
+					mockec2Client.EXPECT().
+						DescribeAddresses(ctx, gomock.AssignableToTypeOf(&ec2.DescribeAddressesInput{})).
+						DoAndReturn(func(_ context.Context, input *ec2.DescribeAddressesInput, _ ...func(*ec2.Options)) (*ec2.DescribeAddressesOutput, error) {
+							switch len(input.Filters) {
+							case 3:
+								Expect(input.Filters).To(ConsistOf(
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyManaged)),
+										Values: []string{"true"},
+									},
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyClusterID)),
+										Values: []string{clusterID},
+									},
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyExternalIPName)),
+										Values: []string{externalIP.Name},
+									},
+								))
+								return &ec2.DescribeAddressesOutput{}, nil
+							case 4:
+								Expect(input.Filters).To(ConsistOf(
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyManaged)),
+										Values: []string{"true"},
+									},
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyClusterID)),
+										Values: []string{clusterID},
+									},
+									types.Filter{
+										Name:   aws.String(fmt.Sprintf("tag:%s", TagKeyExternalIPName)),
+										Values: []string{externalIP.Name},
+									},
+									types.Filter{
+										Name:   aws.String("allocation-id"),
+										Values: []string{allocationID},
+									},
+								))
```

Comment:
I'm not sure, why is there a switch here ? Don't you know how many filters you should get ?